### ISSUE TITLE
Quotation/Invoice template texts not translated

### ISSUE THREAD

**Issue by ramonski** [odoo/odoo#11130]
Impacted versions:
9.saas~9c (Community Edition)

Steps to reproduce:

Within a fresh installation of Odoo

#. Install German Language
#. Create a customer with language=German
#. Create a new Quotation
#. Press Print `Quotation / Order`

Current behavior:

Static strings in the header and footer are not translated into German. Also the table header (`Description`, `Quantity`, `Unit Price`, ...) stay in English.

The log prints the following message:
`2016-02-29 05:55:44,030 36691 INFO ridingbytes openerp.addons.base.ir.ir_qweb: 'doc.partner_id.lang' is not a valid language code, is an empty field or is not installed, falling back to en_US`

Hard coding it to `de_DE` makes no difference.

The described behavior is for standard and custom PDF and HTML reports, which are created according to the documentation:
https://www.odoo.com/documentation/9.0/reference/reports.html

- Installing the Language with `Overwrite Existing Terms` again makes no difference
- Synchronizing Application Terms makes no difference

Expected behavior:

Static strings within the Invoice/Quotaion Templates get translated into the language of the customer.

Video/Screenshot link (optional):

- HTML

![odoo_report](https://cloud.githubusercontent.com/assets/713193/13386917/e1f8f4a4-deb1-11e5-9995-51fde49e367b.png)

- PDF

![quotation_-_order_pdf](https://cloud.githubusercontent.com/assets/713193/13386942/20b4481a-deb2-11e5-9c2e-18ed85fceae8.png)

- Custom Report

![odoo_report](https://cloud.githubusercontent.com/assets/713193/13386983/93c81534-deb2-11e5-8a72-a2f1d2b42269.png)

- All Fields are properly translated

![translate_-_odoo](https://cloud.githubusercontent.com/assets/713193/13387028/f84e32e0-deb2-11e5-8c43-1b993b49c5cd.png)

**Comment by JKE-be** [comment]
Just tested in runbot 9 community... and it works.
[Quotation - Order (1).pdf](https://github.com/odoo/odoo/files/150582/Quotation.-.Order.1.pdf)

But, your warning is strange, can you try to print the lang code in addons/base/ir/ir_qweb.py:render_tag_call.

**Comment by ramonski** [comment]
The problem is in this code segment:
https://github.com/odoo/odoo/blob/saas-9/openerp/addons/base/ir/ir_qweb.py#L445

The `lang` variable used for the domain is the un-evaluated string from the template, which is `doc.partner_id.lang`. The code should do something like this:

```python
    def render_tag_call(self, element, template_attributes, generated_attributes, qwebcontext):
        d = qwebcontext.copy()

        if 'lang' in template_attributes:
            init_lang = d.context.get('lang', 'en_US')
            t_lang = template_attributes['lang'] # returns doc.partner_id.lang
            lang = self.eval(t_lang, d) or init_lang # evaluate the language
            if not self.pool['res.lang'].search(d.cr, d.uid, [('code', '=', lang)], count=True, context=d.context):
                _logger.info("'%s' is not a valid language code, is an empty field or is not installed, falling back to en_US", lang)
```

**Comment by ramonski** [comment]
But I still don't get the translated report:(

I was able to translate the terms within this method using the `openerp.tools.translate`: 
https://github.com/odoo/odoo/blob/saas-9/openerp/addons/base/ir/ir_qweb.py#L349

like this:

```python
    def render_text(self, text, element, qwebcontext):
        context = qwebcontext
        text = _(text)
        return text.encode('utf-8')
```

The company is set to German language and even the Admin user has German language set in the preferences. Any ideas? I saw your report contains `$` as currency. Might this be an issue with `l10n_de_skr03`?

**Comment by ramonski** [comment]
Did the same on a fresh DB loaded with Demo Data and installed Sale Module and loaded German translation:

1. Set customer to German:
![agrolait_-_odoo](https://cloud.githubusercontent.com/assets/713193/13391508/1eec62e2-ded5-11e5-9202-3e4fbbd542c8.png)

2. New Quotation for this customer:
![so008_-_odoo](https://cloud.githubusercontent.com/assets/713193/13391528/3e14d0d2-ded5-11e5-8b9b-6b5cf700a23e.png)

3. Print (HTML)
![odoo_report](https://cloud.githubusercontent.com/assets/713193/13391554/5fc0ba2a-ded5-11e5-85bd-ebf87c0f52c7.png)

4. Print (PDF)

[Quotation - Order (1).pdf](https://github.com/odoo/odoo/files/150659/Quotation.-.Order.1.pdf)

**Comment by JKE-be** [comment]
can you try to reproduce your bug here:
http://134113-saas-9-b4638a-all.runbot.odoo.com/?db=134113-saas-9-b4638a-all

thanks

**Comment by JKE-be** [comment]
About the warning, it a bug... you will always got it, but don't worry, it's just a warning. that have no impact on the code. we will fix it asap.

**Comment by ramonski** [comment]
Yes, I saw that, thanks. The string don't get's evaluated as I mentioned in https://github.com/odoo/odoo/issues/11130#issuecomment-190122483

The quotation template works fine on http://134113-saas-9-b4638a-all.runbot.odoo.com/?db=134113-saas-9-b4638a-all

[Quotation - Order (2).pdf](https://github.com/odoo/odoo/files/150700/Quotation.-.Order.2.pdf)

I am currently on the `saas-9` branch * b4638a9 - (HEAD -> saas-9, origin/saas-9) [FIX] crm: overdue is only the next action in the past. (7 days ago) <Jeremy Kersten>

On this branch, the described issue exists for me on a new DB with only Sale module installed.

**Comment by JKE-be** [comment]
are you on the Odoo saas (which db name ?) ? Or in local (why not pull the last 7 days ? ) ?

**Comment by ramonski** [comment]
I'm on my local machine and `git pull` returns this commit:

```
commit b4638a91532e924158146a2e49a461890b77f87c
Author: Jeremy Kersten <jke@odoo.com>
Date:   Mon Feb 22 16:48:04 2016 +0100

    [FIX] crm: overdue is only the next action in the past.

    Don't show oppotunities where next action date is today.
    We have already a filter Today for it.
```

I'm on a Postgres 9.5 DB and initialized it like this:
![odoo](https://cloud.githubusercontent.com/assets/713193/13392360/3733b936-deda-11e5-8b2d-f48c17ae61de.png)

![apps_-_odoo](https://cloud.githubusercontent.com/assets/713193/13392395/7a2e4666-deda-11e5-8158-c351d6b53295.png)

Installed *only* the Sale module:

![sales_management_-_odoo](https://cloud.githubusercontent.com/assets/713193/13392433/c70d29ca-deda-11e5-8807-96c0b450b3bf.png)

Added German Language Support:

![dashboard_-_odoo](https://cloud.githubusercontent.com/assets/713193/13392456/f5e40052-deda-11e5-8069-2827900450bf.png)


Then the described Steps from above:

- Added a new Customer with German Language
- Added new Quotation
- Click Print

**Comment by JKE-be** [comment]
Just tested again in local, and it works ... 
Maybe you should add a pdb and try to debug your self what is happening.

If you really found an issue, don't hesitate to re-open an issue and link this one.

I can just fix the warning: #3a125d3. Once runbot is green, I will merge it.

**Comment by ramonski** [comment]
Ok, this is strange. Did you install all packages or only sale?
Would it be possible to send me your local database dump to rb@ridingbytes.com?

**Comment by JKE-be** [comment]
Just sale with command:

./odoo.py -d report1 -db-filter=report1 -i sale
Install "German / Deutsche"
Change lang on partner 'Agrolait'
Create a quotation and print

**Comment by ramonski** [comment]
Ok, really strange. It works with the command you sent me:
`./odoo.py -d report1 --db-filter=report1 -i sale`
When I do just `./odoo.py`, create the DB in the UI window and install the sale package, I have the described issue... Maybe it is because of the `db_template`

**Comment by ramonski** [comment]
All right, I figured it out! This last comment is just for the curious.

I used an `.openerp_serverrc` configuration, with `dev_mode = True` in it.
Commenting it out solved my issue! I still don't know why it is like this, but I'm happy that it works now.

Thanks a lot @JKE-be for your support and sorry for the noise.

**Comment by luchitocornejo** [comment]
Hi, i'm having this same issue but in spanish, would you be so kind to guide me into the solution? where am i suppose to input the command ./odoo.py -d report1 --db-filter=report1 -i sale?. I'm running odoo 9 on a localhost in a mac. thanks!

**Comment by ramonski** [comment]
Hi @luchitocornejo, try to set `debug_mode = False` in your config. This issue seems to happen only in debug mode.

**Comment by luchitocornejo** [comment]
thank you @ramonski, where exactly in my config? i'm running it on bitnami

**Comment by ramonski** [comment]
Same issue appears between the commits 7c177d91432bc2499b81ff361434021ca079b9c7 and d38d0be995b1eed4d1b03ea6d3c64c331fc94142

I am not creating a new issue, since it is exactly the behavior described.

@JKE-be: Your help is much appreciated

Thanks

## Translation working in this commit

```
commit 7c177d91432bc2499b81ff361434021ca079b9c7
Author: Olivier Dony <odo@openerp.com>
Date:   Fri Apr 29 11:00:10 2016 +0200

    [FIX] mail queue: wrong condition for test mode

    Rev. 56cec604e6313c49582bbd9801686328696b2675 modifies
    the auto-commit behavior of the mail queue scheduler,
    to allow using it within tests.

    The condition was wrong and always True as soon as
    the registry was loaded, permanently disabling
    auto-commit, both during tests and outside tests.

    This broke the duplicate emails protection.

```

## Translation NOT working in this commit

```
commit d38d0be995b1eed4d1b03ea6d3c64c331fc94142
Author: Goffin Simon <sig@odoo.com>
Date:   Tue Jun 28 15:37:46 2016 +0200

    [FIX] product_extended: bom_count appeared twice in a view

    The field bom_count appeared twice in product.template form view and
    then the smart button for "Bill of Materials" always displayed 0 as
    number of BoMs.

    opw:681830
```

**Comment by ramonski** [comment]
## Update

It works if I change the Users language to German – but this is wrong, as it should use the language of the customer instead `doc.partner_id.lang`, which is in my case `de_DE`.

This worked perfectly in commit 56cec604e6313c49582bbd9801686328696b2675 with my current user language set to English. After updating to the latest commit in the `saas-9` branch (d38d0be995b1eed4d1b03ea6d3c64c331fc94142) it stopped working and I had to change my User language to German.

**Comment by mart-e** [comment]
Hello,

Are you running saas-9 on a local instance? I would really not advise to do so as these versions are not supported (you will no longer get commits in saas-9 now, everything is done in saas-10 or 11 and will change again in 2 months).
Can you reproduce it on a saas-10 instance ? I just tested on the runbot and I got my report correctly translated in German.

**Comment by ramonski** [comment]
Hi @mart-e 

Thanks for your quick response.
Yes, I am running saas-9 on a local instance.
I did not test it yet in saas-10, but will do so.

I managed to "hack" a workaround for this issue on my side. I followed the "Custom Reports" instructions on https://www.odoo.com/documentation/9.0/reference/reports.html and added this class to my module:

```python
from openerp import api
from openerp import models


class SaleOrderReport(models.AbstractModel):
    _name = 'report.ridingbytes_reports.report_saleorder'

    @api.multi
    def render_html(self, *args, **kw):

        report_obj = self.env['report']
        report = report_obj._get_report_from_name('ridingbytes_reports.report_saleorder')
        docs = self.env[report.model].browse(self.ids)

        # XXX: hack to fix the report language
        lang = docs.partner_id.lang

        docargs = {
            'doc_ids': self._ids,
            'doc_model': report.model,
            'docs': docs,
        }

        # XXX: hack to fix the report language
        return report_obj.with_context({"lang": lang}).render('ridingbytes_reports.report_saleorder', docargs)
```

I don't have that much experience with the system yet, so I hacked in the language to the context manually before calling `render`.

Your advice is much appreciated.
Thanks

**Comment by mart-e** [comment]
Again I don't think you should use saas-9 (or saas-10 or 11). The problem is that we do not push new commits in this branch (so even if we find out a bug here, it won't be fixed as it's already working in the latest saas). Also you won't get upgrades to upper version.
Stick to stable versions for on premise deployments, you can see saas as "snapshots" to give a preview of new dev but should not use it to develop modules and definitely not for production.

**Comment by ramonski** [comment]
Thanks Martin!
Indeed, I misunderstood the `saas-9`/`saas-X` branch layout a little bit. I thought of it as a more stable version of the `9.0` branch.
I try to migrate my instance to `9.0`, but saw already that it won't be an easy task.
Bests, Ramon

**Comment by mart-e** [comment]
No, this is a way for us to offer a more smooth transitions to our customers on odoo.com between versions (one small release each few months instead of a big upgrade each year) but supporting these properly would require a lot of work and will prevent us to make frequent release.

Hopefully there isn't too much differences between 9.0 and saas-9 so an upgrade of `base` will maybe enough...

**Comment by mart-e** [comment]
And by the way, just found the reason. e1deae7eab8be77d7bcb46a68560b86ffbb2277a introduced the bug and was reverted at 7f2e748a2245838d09d7f3f17f5e1a1b202f6076. The problem is that the first commit is in saas-9 and saas-10 was released before the revert.

**Comment by ramonski** [comment]
Hi Martin

Thank you very much for the "upgrade hint".

I just tested it on my local machine and migrated to `9.0` by just upgrading `base` and it worked!

I understand now the git branch layout much better, so thanks for taking your time to explain it to me as well.

Cool that you found the issue that caused this bug – besides all the noise I created with this issue, I hope it will help to improve the code for future versions as well.

Thanks an have a nice day
Ramon
